src/body/response/body.rs

```diff
@@ -37,83 +38,84 @@ enum BodyKind {
 
 /// A projection type of the `BodyKind` enum that allows for the inner kind to
 /// participate in API calls that require a pinned reference.
-enum BodyKindProjection<'a> {
+enum ResponseBodyKindProjection<'a> {
     Buffer(Pin<&'a mut BytesMut>),
     Stream(Pin<&'a mut DynStream>),
 }
 
-/// A stream adapter that converts a stream of `Result<D, E>` into a stream of
-/// `Result<Frame<Bytes>>`. This adapter allows for response bodies to be built
-/// from virtually any stream that yields data that can be converted into bytes.
-#[must_use = "streams do nothing unless polled"]
-struct BodyStreamAdapter<T, D, E>
-where
-    T: Stream<Item = Result<D, E>> + Send,
-    Bytes: From<D>,
-    Error: From<E>,
-{
-    /// The `Stream` that we are adapting to yield `Result<Frame<Bytes>>`.
-    stream: T,
-
-    /// This field is used to mark `BodyStreamAdapter` as `!Unpin`. This is
-    /// necessary because `T` may not be `Unpin` and we need to project the
-    /// `stream` through a pinned reference to `Self` so it can be polled.
-    _pin: PhantomPinned,
-}
+impl ResponseBody {
+    pub fn is_empty(&self) -> bool {
+        self.len().map_or(false, |len| len == 0)
+    }
 
-impl Body {
     pub fn len(&self) -> Option<usize> {
         self.as_buffer().map(BytesMut::len)
     }
-
-    pub fn is_empty(&self) -> bool {
-        self.len().map_or(false, |len| len == 0)
-    }
 }
 
-impl Body {
-    pub(super) fn new() -> Self {
+impl ResponseBody {
+    pub(crate) fn new() -> Self {
         let buffer = Box::new(BytesMut::new());
 
         Self {
-            kind: BodyKind::Buffer(buffer),
+            kind: ResponseBodyKind::Buffer(buffer),
             _pin: PhantomPinned,
         }
     }
 
-    pub(super) fn buffer(bytes: Bytes) -> Self {
+    pub(crate) fn buffer(bytes: Bytes) -> Self {
         let buffer = Box::new(BytesMut::from(bytes));
 
         Self {
-            kind: BodyKind::Buffer(buffer),
+            kind: ResponseBodyKind::Buffer(buffer),
             _pin: PhantomPinned,
         }
     }
 
-    pub(super) fn stream<T, D: 'static, E: 'static>(stream: T) -> Self
+    pub(crate) fn stream<T, D: 'static, E: 'static>(stream: T) -> Self
     where
         T: Stream<Item = Result<D, E>> + Send + 'static,
         Bytes: From<D>,
         Error: From<E>,
     {
-        let stream = Box::new(BodyStreamAdapter::new(stream));
+        let stream = Box::new(StreamAdapter::new(stream));
 
         Self {
-            kind: BodyKind::Stream(stream),
+            kind: ResponseBodyKind::Stream(stream),
             _pin: PhantomPinned,
         }
     }
 
+    // TODO:
+    // Determine if there is a way we can compose closures of nested map
+    // functions to avoid recursive pointers as the value of `BodyKind::Stream`.
+    //
+    // Perhaps we can use a `Box<dyn Fn(Bytes) -> Result<Bytes, E> + Send>` to
+    // store the map functions and unwind them in the `poll_frame` method.
+    pub(crate) fn map<F, E: 'static>(self, _: F) -> Self
```

Comment:
I think the TODO comment is a cool idea but would like some feedback with regards to security and performance with this approach. It seems a bit novel but I like the idea of making a slight compromise here if it makes the code easier to reason about / work with. I also have a feeling the performance may be similar to a statically dispatched version that uses recursive concrete types.